src/middleware/devtools.ts

```diff
@@ -17,326 +18,212 @@ type DevtoolsType = {
-            case 'JUMP_TO_ACTION':
-              return parseJsonThen<S>(message.state, (state) => {
-                setStateFromDevtools(state)
-              })
+  const initialState = fn(api.setState, get, api)
+  extension.init(initialState)
 
-            case 'IMPORT_STATE': {
-              const { nextLiftedState } = message.payload
-              const lastComputedState =
-                nextLiftedState.computedStates.slice(-1)[0]?.state
-              if (!lastComputedState) return
-              setStateFromDevtools(lastComputedState)
-              extension.send(null, nextLiftedState)
+  extension.subscribe((message: any) => {
+    switch (message.type) {
+      case 'ACTION':
+        if (typeof message.payload !== 'string') {
+          console.error(
+            '[zustand devtools middleware] Unsupported action format'
+          )
+          return
+        }
+        return parseJsonThen<{ type: unknown; state?: PartialState<S> }>(
+          message.payload,
+          (action) => {
+            if (action.type === '__setState') {
+              setStateFromDevtools(action.state as PartialState<S>)
               return
             }
 
-            case 'PAUSE_RECORDING':
-              return (isRecording = !isRecording)
+            if (!(api as any).dispatchFromDevtools) return
+            if (typeof (api as any).dispatch !== 'function') return
+            ;(api as any).dispatch(action)
           }
-          return
+        )
+
+      case 'DISPATCH':
+        switch (message.payload.type) {
+          case 'RESET':
+            setStateFromDevtools(initialState)
+            return extension.init(api.getState())
+
+          case 'COMMIT':
+            return extension.init(api.getState())
+
+          case 'ROLLBACK':
+            return parseJsonThen<S>(message.state, (state) => {
+              setStateFromDevtools(state)
+              extension.init(api.getState())
+            })
+
+          case 'JUMP_TO_STATE':
+          case 'JUMP_TO_ACTION':
+            return parseJsonThen<S>(message.state, (state) => {
+              setStateFromDevtools(state)
+            })
+
+          case 'IMPORT_STATE': {
+            const { nextLiftedState } = message.payload
+            const lastComputedState =
+              nextLiftedState.computedStates.slice(-1)[0]?.state
+            if (!lastComputedState) return
+            setStateFromDevtools(lastComputedState)
+            extension.send(null, nextLiftedState)
+            return
+          }
+
+          case 'PAUSE_RECORDING':
+            return (isRecording = !isRecording)
+        }
+        return
+    }
+  })
+
+  if (
+    (api as any).dispatchFromDevtools &&
+    typeof (api as any).dispatch === 'function'
+  ) {
+    let didWarnAboutReservedActionType = false
+    const originalDispatch = ((api as any).dispatch(api as any).dispatch = (
+      ...a: any[]
+    ) => {
+      if (a[0].type === '__setState' && !didWarnAboutReservedActionType) {
```

Comment:
We will be able to use `__DEV__`. We can use it to only warn in the development mode.
Also, I wonder if we can/should eliminate `didWarnAboutReservedActionType` and warn many times.

Reply:
Yeah sorry forgot to address this one, yeah I'm using `__DEV__` now. Warning many times will might make the console too noisy for the user (and hence it might unnecessarily force them to fix this their first priority).